packages/widgets/src/blocks/legacy-widget/edit/preview.js

```diff
@@ -7,57 +7,75 @@ import classnames from 'classnames';
  * WordPress dependencies
  */
 import { useRefEffect } from '@wordpress/compose';
-import { addQueryArgs } from '@wordpress/url';
 import { useState } from '@wordpress/element';
 import { Placeholder, Spinner, Disabled } from '@wordpress/components';
 import { __ } from '@wordpress/i18n';
+import apiFetch from '@wordpress/api-fetch';
+
+async function renderPreview( idBase, instance ) {
+	const restRoute = `/wp/v2/widget-types/${ idBase }/render`;
+	return await apiFetch( {
+		path: restRoute,
+		method: 'POST',
+		data: {
+			id_base: idBase,
+			instance,
+		},
+	} );
+}
 
 export default function Preview( { idBase, instance, isVisible } ) {
 	const [ isLoaded, setIsLoaded ] = useState( false );
 
 	// Resize the iframe on either the load event, or when the iframe becomes visible.
 	const ref = useRefEffect(
 		( iframe ) => {
+			if ( ! iframe.previewLoaded ) {
+				renderPreview( idBase, instance ).then( ( response ) => {
+					iframe.srcdoc = response.preview;
```

Comment:
It's probably better to set state here instead of manipulating DOM directly. We'll have to create a new state and call the setState function here.

```js
const [ srcDoc, setSrcDoc ] = useState( '' );

renderPreview( idBase, instance ).then( ( response ) => {
  setSrcDoc( response.preview );
} );

<iframe srcDoc={ srcDoc } />
```

We'll probably also want to add some error handling and abort the fetch when the component is unmounted too. Which can be a follow-up PR.

Reply:
Also, this request isn't depend on the reference of the iframe node itself, so I don' think it's necessary to put it in `useRefEffect`. Instead, using `useEffect` should be enough.

We should also drop the custom `previewLoaded` field too.

Reply:
> We should also drop the custom previewLoaded field too.
I don't think we need to drop `previewLoaded` as we only need to fetch the preview HTML once.
We need some variable that we can use a as a flag to control if the `apiFetch` function has already been called. We can't use `srcdoc` for that because if the apiFetch request fails, the srcdoc attribute will be empty and we can't use it to find out if we've already called apiFetch.
But I agree that we probably need to create a new state for that.

Reply:
Yep, I mean to `useState` for that rather than adding unsupported properties to DOM elements.

Reply:
@kevin940726 Thank you for the feedback!
I hope I've fixed all the issues you mentioned.
Could you please check it?